### PR TITLE
feat: add support for CoW in Win Dev Drives

### PR DESCRIPTION
**Pull request by nachoaldamav**
Bump the `reflink-copy` dependency version to `0.1.8` as it supports now Copy-on-Write on the new Windows Dev Drives.

Keeping it as draft for now until I can test it in a Dev Drive later today.

Reference: [https://github.com/cargo-bins/reflink-copy/pull/28](https://github.com/cargo-bins/reflink-copy/pull/28)

### CODE REVIEW

**Comment by nachoaldamav** [comment]
Looks like it works, but only in Admin mode, if not in admin mode it throws the following error, I'll check it.

```bash
PS D:\test-pacquet> D:\pacquet\target\debug\pacquet.exe add axios
Error: pacquet_package_manager::io_error

  × adding a new package
  ╰─▶ The client does not have a required privilege (os error 1314).
````

image

File: crates/cli/src/fs.rs [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
-    os::unix::fs::symlink(original, link)
+    std::os::unix::fs::symlink(original, link)
 }
 
 #[cfg(windows)]
 pub fn symlink_dir(original: &Path, link: &Path) -> io::Result<()> {
-    os::windows::fs::symlink_dir(original, link)
+    match std::os::windows::fs::symlink_dir(original, link) {
+        Ok(_) => Ok(()),
+        Err(_) => {
```
Comment (anonrig) [inline crates/cli/src/fs.rs:12]: Would it makes sense to look into the error type in order to create a junction for performance reasons?

Comment (nachoaldamav) [reply crates/cli/src/fs.rs:12]: Something like creating a junction only when we have the elevated privileges error?

Comment (anonrig) [reply crates/cli/src/fs.rs:12]: Yes

Comment (nachoaldamav) [reply crates/cli/src/fs.rs]: I added another commit, now it searchs for the error code `1314`.

**Comment by nachoaldamav** [comment]
@anonrig I've found another issue, in Dev Drives the packages are copied, not cloned, because the store is in the NFTS drive, so the clone function fallback to copy. I'll try to fix it.

**Comment by anonrig** [comment]
Can you add some tests @nachoaldamav?

**Comment by nachoaldamav** [comment]
> Can you add some tests @nachoaldamav?

Yes, I'll try to do it tomorrow

File: crates/cli/src/fs.rs [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+    match std::os::windows::fs::symlink_dir(original, link) {
+        Ok(_) => Ok(()),
+        Err(e) => {
+            match e.raw_os_error() {
+                // If we don't have the privilege to create a symlink, we can try to create a junction instead.
+                Some(ERROR_PRIVILEGE_NOT_HELD) => junction::create(original, link),
+                _ => Err(e),
+            }
+        }
+    }
```
Comment (zkochan) [inline crates/cli/src/fs.rs:22]: Just use junctions on Windows. It is not optimal to always try a symlink and then retry a junction. It will be a lot more file system operations.

File: crates/cli/src/fs.rs [hunk 3]
```diff
@@ -1,13 +1,25 @@
-use std::{io, os, path::Path};
+use std::{io, path::Path};
+
+#[cfg(windows)]
+const ERROR_PRIVILEGE_NOT_HELD: i32 = 1314;
 
 #[cfg(unix)]
 pub fn symlink_dir(original: &Path, link: &Path) -> io::Result<()> {
-    os::unix::fs::symlink(original, link)
+    std::os::unix::fs::symlink(original, link)
```
Comment (zkochan) [inline crates/cli/src/fs.rs:8]: Why is this change needed?

Comment (nachoaldamav) [reply crates/cli/src/fs.rs:8]: Not intented, I'll revert that

File: crates/npmrc/src/custom_deserializer.rs [hunk 4]
```diff
@@ -16,6 +16,18 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
     vec!["*eslint*".to_string(), "*prettier*".to_string()]
 }
 
+// Get the drive letter from a path on Windows. If it's not a Windows path, return None.
+fn get_drive_letter(current_dir: &Path) -> Option<String> {
```
Comment (anonrig) [inline crates/npmrc/src/custom_deserializer.rs:20]: We don't need to iterate through the components if we are not running on Windows.

File: crates/npmrc/src/custom_deserializer.rs [hunk 5]
```diff
@@ -31,16 +43,26 @@ pub fn default_store_dir() -> PathBuf {
...
-    match env::consts::OS {
-        "linux" => home_dir.join(".local/share/pacquet/store"),
-        "macos" => home_dir.join("Library/pacquet/store"),
-        "windows" => home_dir.join("AppData/Local/pacquet/store"),
-        _ => panic!("unsupported operating system: {}", env::consts::OS),
+    // In case of Windows, we need to get the drive letter of the home directory, to use the store folder on the same drive.
+    let home_drive = get_drive_letter(&home_dir).unwrap_or_default();
+
+    // Check if we are on the same drive as the home directory
+    if current_drive == home_drive {
```
Comment (anonrig) [inline crates/npmrc/src/custom_deserializer.rs:56]: This check should only happen on Windows

File: crates/cli/src/fs.rs [hunk 6]
```diff
@@ -7,7 +7,7 @@ pub fn symlink_dir(original: &Path, link: &Path) -> io::Result<()> {
 
 #[cfg(windows)]
 pub fn symlink_dir(original: &Path, link: &Path) -> io::Result<()> {
-    os::windows::fs::symlink_dir(original, link)
+    junction::create(original, link)
```
Comment (zkochan) [inline crates/cli/src/fs.rs:10]: Add a comment about why we use junctions on Windows.

**Comment by nachoaldamav** [comment]
As now we are using just junctions for Windows there is a failing test ([install.rs#L304](https://github.com/pnpm/pacquet/blob/main/crates/cli/src/commands/install.rs#L304))

File: crates/npmrc/src/custom_deserializer.rs [hunk 7]
```diff
@@ -127,4 +154,30 @@ mod tests {
...
+
+    #[cfg(target_os = "windows")]
+    #[test]
+    fn test_should_get_the_correct_drive_letter() {
+        let current_dir = Path::new("C:\\Users\\user\\project");
+        let drive_letter = get_drive_letter(current_dir);
+        assert_eq!(drive_letter, Some("C".to_string()));
+    }
+
+    #[cfg(target_os = "windows")]
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:166]:
```suggestion
    #[cfg(windows)]
```

is sufficient

File: crates/npmrc/src/custom_deserializer.rs [hunk 8]
```diff
@@ -127,4 +154,30 @@ mod tests {
         assert_eq!(store_dir, Path::new("/tmp/xdg_data_home/pacquet/store"));
         env::remove_var("XDG_DATA_HOME");
     }
+
+    #[cfg(target_os = "windows")]
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:158]:
```suggestion
    #[cfg(windows)]
```

this too

File: crates/npmrc/src/custom_deserializer.rs [hunk 9]
```diff
@@ -16,6 +16,30 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
     vec!["*eslint*".to_string(), "*prettier*".to_string()]
 }
 
+// Get the drive letter from a path on Windows. If it's not a Windows path, return None.
+fn get_drive_letter(current_dir: &Path) -> Option<String> {
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:20]: I recommend `Cow<'_, str>`. It would avoid allocation when the path is relative.

File: crates/npmrc/src/custom_deserializer.rs [hunk 10]
```diff
@@ -16,6 +16,30 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
+    for component in current_dir.components() {
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:21]: Drive letter will only ever exist as the very first component, so looping is sub-optimal.

```suggestion
    if let Some(component) = current_dir.components().next() {
```

File: crates/npmrc/src/custom_deserializer.rs [hunk 11]
```diff
@@ -16,6 +16,30 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
+        if let Component::Prefix(prefix_component) = component {
+            return Some(
+                prefix_component.as_os_str().to_str().unwrap().replace(':', "").to_string(),
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:24]: This is error-prone and incomplete. Please use [`prefix_component.kind()`](https://doc.rust-lang.org/stable/std/path/struct.PrefixComponent.html#method.kind).

File: crates/npmrc/src/custom_deserializer.rs [hunk 12]
```diff
@@ -155,12 +155,12 @@ mod tests {
...
     }
 
-    #[cfg(target_os = "windows")]
+    #[cfg(windows)]
     #[test]
     fn test_should_get_the_correct_drive_letter() {
         let current_dir = Path::new("C:\\Users\\user\\project");
         let drive_letter = get_drive_letter(current_dir);
-        assert_eq!(drive_letter, Some("C".to_string()));
+        assert_eq!(drive_letter, Some(Cow::Borrowed("C")));
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:163]:
```suggestion
        assert_eq!(drive_letter.as_deref(), Some("C"));
```

File: crates/npmrc/src/custom_deserializer.rs [hunk 13]
```diff
@@ -17,12 +17,12 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
-        if let Component::Prefix(prefix_component) = component {
-            return Some(
-                prefix_component.as_os_str().to_str().unwrap().replace(':', "").to_string(),
-            );
+fn get_drive_letter(current_dir: &Path) -> Option<Cow<'_, str>> {
+    if let Some(Component::Prefix(prefix_component)) = current_dir.components().next() {
+        if let std::path::Prefix::Disk(disk_byte) | std::path::Prefix::VerbatimDisk(disk_byte) =
+            prefix_component.kind()
+        {
+            return Some(Cow::Owned((disk_byte as char).to_string()));
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:25]: Actually, why return `Option<Cow<str>>` in the first place? Why not just return `Option<char>`? Drive letters don't need a whole string.

File: crates/cli/src/commands/install.rs [hunk 14]
```diff
@@ -218,6 +219,19 @@ mod tests {
...
+    fn is_symlink_or_junction(path: std::path::PathBuf) -> Result<bool> {
+        #[cfg(windows)]
+        {
+            junction::exists(&path)
+        }
+
+        #[cfg(not(windows))]
+        {
+            Ok(path.is_symlink())
+        }
```
Comment (KSXGitHub) [inline crates/cli/src/commands/install.rs:232]: `return` statements use less lines:

```suggestion
        #[cfg(unix)]
        return Ok(path.is_symlink());
        #[cfg(windows)]
        return junction::exists(&path);
```

File: crates/npmrc/src/custom_deserializer.rs [hunk 15]
```diff
@@ -16,6 +15,30 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
+            return Some(disk_byte as char);
+        }
+    }
+    None
+}
+
+fn default_store_dir_windows(home_dir: PathBuf) -> PathBuf {
+    let current_dir = env::current_dir().expect("Current directory is not available");
+    let current_drive = get_drive_letter(&current_dir).unwrap_or_default();
+    let home_drive = get_drive_letter(&home_dir).unwrap_or_default();
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:33]: `unwrap_or_default` doesn't make sense for a drive letter. I think you should handle the `None` path a bit differently.

Comment (KSXGitHub) [reply crates/npmrc/src/custom_deserializer.rs:33]: Just panic on `None` is fine for now. This project is still WIP after all.

```suggestion
    let current_drive = get_drive_letter(&current_dir).expect("current dir is an absolute path with drive letter");
    let home_drive = get_drive_letter(&home_dir).expect("home dir is an absolute path with drive letter");
```

File: crates/npmrc/src/custom_deserializer.rs [hunk 16]
```diff
@@ -127,4 +161,30 @@ mod tests {
...
+        let current_drive = get_drive_letter(&current_dir).unwrap_or_default();
+        let home_drive = get_drive_letter(&home_dir).unwrap_or_default();
+        let store_drive = get_drive_letter(&store_dir).unwrap_or_default();
+
+        if current_drive == home_drive {
+            assert_eq!(store_drive, home_drive);
+        } else {
+            assert_eq!(store_drive, current_drive);
+        }
+    }
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:189]: This test is essentially the exact duplicate of the tested function, making it an unnecessary test.

The assertions is in the `if` branches, making it inconsistent, unpredictable, and unreliable.

Suggestions:
1. Instead of testing `default_store_dir`, please test `default_store_dir_windows` instead. It would allow passing the `home_dir` as parameter independent from the hardware and compare the result against expected constant.
2. The `unwrap_or_default` is unnecessary here. You can compare `Option` to `Option`.

Comment (nachoaldamav) [reply crates/npmrc/src/custom_deserializer.rs:189]: What do you think about this? I added a new param to the store dir windows function to pass the current dir, so I can mock it in the tests
```rust
#[cfg(windows)]
#[test]
fn test_default_store_dir_with_windows() {
    let mocked_current_dir = Path::new("D:\\Users\\user\\project");
    let mocked_home_dir = Path::new("C:\\Users\\user");

    let store_dir = default_store_dir_windows(mocked_home_dir.to_path_buf(), mocked_current_dir.to_path_buf());

    let current_drive = get_drive_letter(mocked_current_dir);
    let store_drive = get_drive_letter(&store_dir);

    assert_eq!(current_drive, Some(store_drive.unwrap()));
}
```

So in this specific scenario, as the current directory is in `D`, the store dir should also be in `D`, even if the home dir is in another drive.

Comment (KSXGitHub) [reply crates/npmrc/src/custom_deserializer.rs]: Damn, How could I miss that `default_store_dir_windows` called `current_dir`? But anyway, yeah, this is the approach I prefer. You don't need to prefix the variable with `mocked_` though.

BTW, I have even gone as far as to extract a piece of complex code into a pure function for the purpose of testing: https://github.com/pnpm/pacquet/blob/417581ddd75701630a07160b1cd061ba6409fab9/crates/cli/src/package_manager.rs#L80-L91

File: crates/npmrc/src/custom_deserializer.rs [hunk 17]
```diff
@@ -16,6 +18,33 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
+            prefix_component.kind()
+        {
+            return Some(disk_byte as char);
+        }
+    }
+    None
+}
+
+#[cfg(windows)]
+fn default_store_dir_windows(home_dir: PathBuf, current_dir: PathBuf) -> PathBuf {
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:35]:
```suggestion
fn default_store_dir_windows(home_dir: &Path, current_dir: &Path) -> PathBuf {
```

see https://github.com/pnpm/pacquet/blob/main/CODE_STYLE_GUIDE.md#when-to-use-owned-parameter-when-to-use-borrowed-parameter

File: crates/npmrc/src/custom_deserializer.rs [hunk 18]
```diff
@@ -16,6 +18,33 @@ pub fn default_public_hoist_pattern() -> Vec<String> {
...
+    let current_drive =
+        get_drive_letter(&current_dir).expect("current dir is an absolute path with drive letter");
+    let home_drive =
+        get_drive_letter(&home_dir).expect("home dir is an absolute path with drive letter");
+
+    if current_drive == home_drive {
+        return home_dir.join("AppData/Local/pacquet/store");
+    }
+
+    PathBuf::from(format!("{}:\\.pacquet-store", current_drive))
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:45]:
```suggestion
    PathBuf::from(format!("{current_drive}:\\.pacquet-store"))
```

File: crates/npmrc/src/custom_deserializer.rs [hunk 19]
```diff
@@ -127,4 +161,26 @@ mod tests {
...
+        let current_dir = Path::new("C:\\Users\\user\\project");
+        let drive_letter = get_drive_letter(current_dir);
+        assert_eq!(drive_letter, Some('C'));
+    }
+
+    #[cfg(windows)]
+    #[test]
+    fn test_default_store_dir_with_windows() {
+        let mocked_current_dir = Path::new("D:\\Users\\user\\project");
+        let mocked_home_dir = Path::new("C:\\Users\\user");
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:177]: `mocked_` prefix is unnecessary

File: crates/npmrc/src/custom_deserializer.rs [hunk 20]
```diff
@@ -127,4 +161,26 @@ mod tests {
...
+
+        let store_dir = default_store_dir_windows(mocked_home_dir.to_path_buf(), mocked_current_dir.to_path_buf());
+
+        let current_drive = get_drive_letter(mocked_current_dir);
+        let store_drive = get_drive_letter(&store_dir);
+
+        assert_eq!(current_drive, Some(store_drive.unwrap()));
```
Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:184]:
```suggestion
        assert_eq!(current_drive, store_drive);
```

Comment (KSXGitHub) [inline crates/npmrc/src/custom_deserializer.rs:184]: Why are you testing just the drive letter?

```suggestion

        let store_dir = default_store_dir_windows(mocked_home_dir.to_path_buf(), mocked_current_dir.to_path_buf());
        assert_eq!(&store_dir, Path::new("D:\\.pacquet-store"));
```

Comment (nachoaldamav) [reply crates/npmrc/src/custom_deserializer.rs]: I focused the test just in having the correct drive, but yeah, I just changed that

**Review by KSXGitHub — APPROVED** [review]

File: crates/npmrc/src/custom_deserializer.rs [hunk 21]
```diff
@@ -127,4 +161,22 @@ mod tests {
...
+
+    #[cfg(windows)]
+    #[test]
+    fn test_default_store_dir_with_windows() {
+        let current_dir = Path::new("D:\\Users\\user\\project");
+        let home_dir = Path::new("C:\\Users\\user");
+
+        let store_dir = default_store_dir_windows(&home_dir, &current_dir);
+        assert_eq!(store_dir, Path::new("D:\\.pacquet-store"));
+    }
```
Comment (zkochan) [inline crates/npmrc/src/custom_deserializer.rs:181]: i think we need one more test for when the home dir is on the same drive as the current dir.

**Review by zkochan — APPROVED** [review]
